auth/src/main/java/org/aerogear/mobile/security/AsyncSecurityCheckExecutor.java

```diff
@@ -0,0 +1,64 @@
+package org.aerogear.mobile.security;
+
+import android.content.Context;
+
+import org.aerogear.mobile.auth.Callback;
+import org.aerogear.mobile.core.metrics.MetricsService;
+import org.aerogear.mobile.security.metrics.SecurityCheckResultMetric;
+
+import java.util.Collection;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+public class AsyncSecurityCheckExecutor extends AbstractSecurityCheckExecutor {
+
+    private final ExecutorService executorService;
+
+    AsyncSecurityCheckExecutor(final Context context, final ExecutorService executorService, final Collection<SecurityCheck> checks, final MetricsService metricsService) {
+        super(context, checks, metricsService);
+        this.executorService = executorService;
+    }
+
+    public Future<SecurityCheckResult>[] execute() {
+
+        final Collection<SecurityCheck> checks = getChecks();
+        final MetricsService metricsService = getMetricsService();
+
+        final Future[] res = new Future[checks.size()];
+
+        int i = 0;
+        for (final SecurityCheck check : checks) {
+            res[i++] = (executorService.submit(() -> {
+                final SecurityCheckResult result =  check.test(getContext());
+                if (metricsService != null) {
+                    metricsService.publish(new SecurityCheckResultMetric(result));
+                }
+                return result;
+            }));
+        }
+
+        return res;
+    }
+
+    public void execute(final Callback<SecurityCheckResult> callback) {
```

Comment:
is this method really needed?

Reply:
I added this in case the user wants to receive the results as they get produced instead of waiting on the future. We can of course remove it, but I think it can be useful, especially for UI.

Reply:
I can see your point, given how quickly those tests will run, I honest don't think this is necessary. Plus it actually make things a little bit tricker for developers to consume: we need to tell developers that this callback will be invoked every time when a test is finished, and they can't control the order, and they can't know when all the tests finished running easily.

If we are going to add callback for each test, at least add another callback when all the tests finished running.

But as I said, I think this is overkill and over-engineered.

Reply:
> we need to tell developers that this callback will be invoked every time when a test is finished

I actually like this, and pretty much matches also the world of reactive programming, for each 'item' an onNext() is invoked

> If we are going to add callback for each test, at least add another callback when all the tests finished running.

we could do something like "onComplete", similar to the pattern above 


> I think this is overkill and over-engineered.

do you mean the async handling? or the new class structure?